ssl.go

```diff
@@ -95,16 +99,72 @@ func sslLoadCert(ctx *openssl.Ctx, certFile string) (err error) {
 	return
 }
 
-func sslLoadKey(ctx *openssl.Ctx, keyFile string) (err error) {
+func sslLoadKey(ctx *openssl.Ctx, keyFile string, password string,
+	passwordFile string) error {
 	var keyBytes []byte
+	var err error
+
 	if keyBytes, err = ioutil.ReadFile(keyFile); err != nil {
-		return
+		return err
 	}
 
 	var key openssl.PrivateKey
-	if key, err = openssl.LoadPrivateKeyFromPEM(keyBytes); err != nil {
-		return
+	var errs []error
+
+	// If the key is encrypted and password is not provided,
+	// openssl.LoadPrivateKeyFromPEM(keyBytes) asks to enter PEM pass phrase
+	// interactively. On the other hand,
+	// openssl.LoadPrivateKeyFromPEMWithPassword(keyBytes, password) works fine
+	// for non-encrypted key with any password, including empty string. If
+	// the key is encrypted, we fast fail with password error instead of
+	// requesting the pass phrase interactively.
+	key, err = openssl.LoadPrivateKeyFromPEMWithPassword(keyBytes, password)
+	if err == nil {
+		return ctx.UsePrivateKey(key)
+	} else {
+		errs = append(errs, err)
+	}
+
+	if passwordFile != "" {
+		var file *os.File
+		file, err = os.Open(passwordFile)
+		if err == nil {
+			defer file.Close()
+
+			scanner := bufio.NewScanner(file)
+			// Tarantool itself tries each password file line.
+			for scanner.Scan() {
+				password = strings.TrimSpace(scanner.Text())
+
+				key, err = openssl.LoadPrivateKeyFromPEMWithPassword(keyBytes, password)
+				if err == nil {
+					return ctx.UsePrivateKey(key)
+				} else {
+					errs = append(errs, err)
+				}
+			}
+		} else {
+			errs = append(errs, err)
+		}
+	}
```

Comment:
Let's rewrite this code a bit:

```Go
passwords := []string{password}
if passwordFile != "" {
	// Read lines from a password file.
	passwords = append(passwords, lines...)
}

for _, password := range password {
	key, err = openssl.LoadPrivateKeyFromPEMWithPassword(keyBytes, password)
	if err == nil {
		return ctx.UsePrivateKey(key)
	} else {
		errs = append(errs, err)
	}
}
```

It seems to me that it would be clearer.

Reply:
Thank you, updated